File: src/com/maddyhome/idea/vim/action/change/OperatorAction.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2016 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.action.change;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.maddyhome.idea.vim.KeyHandler;
import com.maddyhome.idea.vim.VimPlugin;
import com.maddyhome.idea.vim.action.VimCommandAction;
import com.maddyhome.idea.vim.command.*;
import com.maddyhome.idea.vim.common.TextRange;
import com.maddyhome.idea.vim.group.MotionGroup;
import com.maddyhome.idea.vim.handler.EditorActionHandlerBase;
import com.maddyhome.idea.vim.helper.MessageHelper;
import com.maddyhome.idea.vim.key.OperatorFunction;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.util.List;
import java.util.Set;

/**
 * @author vlan
 */
public class OperatorAction extends VimCommandAction {
  protected OperatorAction() {
    super(new EditorActionHandlerBase() {
      @Override
      protected boolean execute(@NotNull Editor editor, @NotNull DataContext context, @NotNull Command cmd) {
        final OperatorFunction operatorFunction = VimPlugin.getKey().getOperatorFunction();
        if (operatorFunction != null) {
          final Argument argument = cmd.getArgument();
          if (argument != null) {
            final Command motion = argument.getMotion();
            if (motion != null) {
              final TextRange range = MotionGroup
                .getMotionRange(editor, editor.getCaretModel().getPrimaryCaret(), context, cmd.getCount(),
                                cmd.getRawCount(), argument, true);
              if (range != null) {
                VimPlugin.getMark().setChangeMarks(editor, range);
                final SelectionType selectionType = SelectionType.fromCommandFlags(motion.getFlags());
                KeyHandler.getInstance().reset(editor);
                return operatorFunction.apply(editor, context, selectionType);
              }
            }
          }
          return false;
        }
        VimPlugin.showMessage(MessageHelper.message("E774"));
        return false;
      }
    });
  }

  @NotNull
  @Override
  public Set<MappingMode> getMappingModes() {
    return MappingMode.N;
  }

  @NotNull
  @Override
  public Set<List<KeyStroke>> getKeyStrokesSet() {
    return parseKeysSet("g@");
  }

  @NotNull
  @Override
  public Command.Type getType() {
    return Command.Type.OTHER_SELF_SYNCHRONIZED;
  }

  @NotNull
  @Override
  public Argument.Type getArgumentType() {
    return Argument.Type.MOTION;
  }

  @Override
  public int getFlags() {
    return Command.FLAG_OP_PEND;
  }
}

File: src/main/java/com/maddyhome/idea/vim/action/change/OperatorAction.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.action.change

import com.intellij.vim.annotations.CommandOrMotion
import com.intellij.vim.annotations.Mode
import com.maddyhome.idea.vim.KeyHandler
import com.maddyhome.idea.vim.api.ExecutionContext
import com.maddyhome.idea.vim.api.Options
import com.maddyhome.idea.vim.api.VimCaret
import com.maddyhome.idea.vim.api.VimEditor
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.api.setChangeMarks
import com.maddyhome.idea.vim.command.Argument
import com.maddyhome.idea.vim.command.Command
import com.maddyhome.idea.vim.command.OperatorArguments
import com.maddyhome.idea.vim.common.TextRange
import com.maddyhome.idea.vim.common.argumentCaptured
import com.maddyhome.idea.vim.ex.ExException
import com.maddyhome.idea.vim.group.MotionGroup
import com.maddyhome.idea.vim.group.visual.VimSelection
import com.maddyhome.idea.vim.handler.VimActionHandler
import com.maddyhome.idea.vim.handler.VisualOperatorActionHandler
import com.maddyhome.idea.vim.helper.inRepeatMode
import com.maddyhome.idea.vim.newapi.ij
import com.maddyhome.idea.vim.options.OptionAccessScope
import com.maddyhome.idea.vim.state.mode.SelectionType
import com.maddyhome.idea.vim.vimscript.model.CommandLineVimLContext
import com.maddyhome.idea.vim.vimscript.model.expressions.SimpleExpression
import com.maddyhome.idea.vim.vimscript.model.functions.toVimFuncref

// todo make it multicaret
private fun doOperatorAction(
  editor: VimEditor,
  context: ExecutionContext,
  textRange: TextRange,
  motionType: SelectionType,
): Boolean {
  val operatorfunc = injector.optionGroup.getOptionValue(Options.operatorfunc, OptionAccessScope.GLOBAL(editor))
  if (operatorfunc.value.isEmpty()) {
    injector.messages.showStatusBarMessage(editor, injector.messages.message("E774"))
    return false
  }

  val scriptContext = CommandLineVimLContext

  try {
    val funcref = operatorfunc.toVimFuncref(editor, context, scriptContext)

    val arg = when (motionType) {
      SelectionType.LINE_WISE -> "line"
      SelectionType.CHARACTER_WISE -> "char"
      SelectionType.BLOCK_WISE -> "block"
    }

    val saveRepeatHandler = VimRepeater.repeatHandler
    injector.markService.setChangeMarks(editor.primaryCaret(), textRange)
    KeyHandler.getInstance().reset(editor)

    val arguments = listOf(SimpleExpression(arg))
    funcref.execute(arguments, range = null, editor, context, scriptContext)

    VimRepeater.repeatHandler = saveRepeatHandler
    return true
  } catch (e: ExException) {
    injector.messages.showStatusBarMessage(editor, e.message)
    return false
  }
}

@CommandOrMotion(keys = ["g@"], modes = [Mode.NORMAL])
internal class OperatorAction : VimActionHandler.SingleExecution() {
  override val type: Command.Type = Command.Type.OTHER_SELF_SYNCHRONIZED

  override val argumentType: Argument.Type = Argument.Type.MOTION

  override fun execute(
    editor: VimEditor,
    context: ExecutionContext,
    cmd: Command,
    operatorArguments: OperatorArguments,
  ): Boolean {
    val argument = cmd.argument as? Argument.Motion ?: return false
    if (!editor.inRepeatMode) {
      argumentCaptured = argument
    }
    val range = getMotionRange(editor, context, argument, operatorArguments)
    if (range != null) {
      return doOperatorAction(editor, context, range, argument.getMotionType())
    }
    return false
  }

  private fun getMotionRange(
    editor: VimEditor,
    context: ExecutionContext,
    argument: Argument.Motion,
    operatorArguments: OperatorArguments,
  ): TextRange? {
    // Note that we're using getMotionRange2 in order to avoid normalising the linewise range into line start
    // offsets that will be used to set the change marks. This affects things like the location of the caret in the
    // Commentary extension
    val ijEditor = editor.ij
    return MotionGroup.getMotionRange2(
      ijEditor,
      ijEditor.caretModel.primaryCaret,
      context.ij,
      argument,
      operatorArguments,
    )?.normalize()?.let {
      // If we're linewise, make sure the end offset isn't just the EOL char
      if (argument.getMotionType() == SelectionType.LINE_WISE && it.endOffset < editor.fileSize()) {
        TextRange(it.startOffset, it.endOffset + 1)
      } else {
        it
      }
    }
  }
}

@CommandOrMotion(keys = ["g@"], modes = [Mode.VISUAL])
internal class VisualOperatorAction : VisualOperatorActionHandler.ForEachCaret() {
  override val type: Command.Type = Command.Type.OTHER_SELF_SYNCHRONIZED

  override fun executeAction(
    editor: VimEditor,
    caret: VimCaret,
    context: ExecutionContext,
    cmd: Command,
    range: VimSelection,
    operatorArguments: OperatorArguments,
  ): Boolean {
    return doOperatorAction(editor, context, range.toVimTextRange(), range.type)
  }
}
